File: 2016-2/b.py

from itertools import combinations
import sys


def product(l):
    p = 1.0
    for x in l:
        p *= x
    return p


def ptie(probs):
    ans = 0.0
    halfs = list(combinations(probs, k/2))
    for yes, no in zip(halfs, halfs[::-1]):
        ans += product(yes) * product(1.0-p for p in no)
    return ans


def brute(probs, k):
    ans = 0.0
    best = None
    for subset in combinations(probs, k):
        p = ptie(subset)
        if p > ans:
            ans = p
            best = subset
    print best
    return ans


def greedy(probs, k):
    probs = sorted(probs)
    ans = 0.0
    for i in range(k+1):
        subset = probs[:i] + (probs[-(k-i):] if i<k else [])
        ans = max(ans, ptie(subset))
    return ans


for tc in range(int(raw_input())):
    _, k = map(int, raw_input().split())
    probs = map(float, raw_input().split())
    # ans = brute(probs, k)
    ans = greedy(probs, k)
    sys.stdout.write("Case #{}: {}\n".format(tc+1, ans))

File: 2016-2/b.cpp

#include <iostream>

#define REP(i,N) for (int i = 0; i < N; i++)

using namespace std;


int T, n, k;
double p[400];


int bitsum(int s) {
    if (s == 0) return 0;
    return 1 + bitsum(s & (s-1));
}


double brute() {
    double ans = 0.0;
    REP(subset, 1<<n) if (bitsum(subset) == k) {
        double ptie = 0.0;
        REP(yes, 1<<n) if ((yes & subset) == yes && bitsum(yes) == k/2) {
            double x = 1.0;
            REP(i, n) if ((subset & (1<<i)) > 0) {
                if ((yes & (1<<i)) > 0) x *= p[i];
                else x *= 1.0 - p[i];
            }
            ptie += x;
        }
        ans = max(ans, ptie);
    }
    return ans;
}


int main() {
    cin >> T;
    REP(tc, T) {
        cin >> n >> k;
        REP(i, n) cin >> p[i];
        double ans = brute();
        cout << "Case #" << tc+1 << ": " << ans << endl;        
    }
    return 0;
}

